bluemira/power_cycle/coilsupply.py

```diff
@@ -0,0 +1,705 @@
+    make_dataclass,
+    replace,
+)
+from typing import Any, Dict, List, Optional, Tuple, Type, Union
+
+import numpy as np
+
+from bluemira.base.look_and_feel import bluemira_print
+from bluemira.power_cycle.errors import CoilSupplySystemError
+from bluemira.power_cycle.net import (
+    Config,
+    Descriptor,
+)
+from bluemira.power_cycle.tools import pp
+
+
+def _get_module_class_from_str(class_name):
+    return getattr(sys.modules[__name__], class_name)
+
+
+class CoilSupplyABC(ABC):
+    """Abstract base class for coil supply systems."""
+
+    @abstractmethod
+    def _just_to_stop_ruff_checks(self):
+        """Define dummy method to stop ruff checks."""
+
+
+@dataclass
+class CoilSupplyParameterABC:
+    """
+    Specifier of parameters for a 'CoilSupplySystem' instance.
+
+    Upon creation of a 'CoilSupplyInputs' instance, this class is used
+    to specify the structure of parameters applied to methods of the
+    'CoilSupplySystem' instance created with the 'CoilSupplyInputs'
+    instance.
+    """
+
+    subclass_name = "CoilSupplyParameter"
+    single_value_types = (bool, int, float, list, tuple, np.ndarray)
+
+    @classmethod
+    def validate_parameter(cls, obj):
+        """
+        Validate 'obj' to be of the same class as the object testing it.
+
+        Substitute method for 'isinstance', since it fails in this
+        implementation.
+        """
+        return obj.__class__.__name__ == cls.__name__
+
+    @classmethod
+    def init_subclass(cls, argument: Any = None):
+        """
+        Create a 'CoilSupplyParameter' subclass instance from argument.
+
+        If 'None' is given to instantiate the class, an empty instance
+        is created.
+        If an object of this class is given to instantiate the class, it
+        is returned as is.
+        If a 'dict' is given to instantiate the class, keys must match
+        class attributes and their values are distributed.
+        If a value of one of the 'single_value_types' classes is given
+        to instantiate the class, copy that value to all attributes.
+        """
+        if argument is None:
+            return cls()
+        if cls.validate_parameter(argument):
+            return argument
+        if isinstance(argument, dict):
+            return cls(**argument)
+        if isinstance(argument, cls.single_value_types):
+            args = {}
+            all_fields = fields(cls)
+            for one_field in all_fields:
+                args[one_field.name] = argument
+            return cls(**args)
+        raise ValueError(
+            "A 'CoilSupplyParameter' instance must be initialized "
+            f"with 'None' for an empty instance, a '{cls.__name__}' "
+            "instance for no alteration, a 'dict' for a distributed "
+            "instantiation or any of the following types for a "
+            f"single-value instantiation: {cls.single_value_types}. "
+            f"Argument was '{type(argument)}' instead."
+        )
```

Comment:
None of this is needed please remove, its over complicating things again
```suggestion
```

Reply:
Hi @je-cook,

this one is a bit more complicated than that... I have made a "dataclass factory" to create dataclasses to store vectors for a `CoilSupplySystem` object. This ensures the dataclass has one vector for each coil defined in the `CoilSupplySystem`. These methods are not so much for validation, but rather automatized creation of these containers. I call the related methods "validation" but they are used to instantiate the container objects independent of the input so maybe the term is not the most appropriate.

In short, the workflow is as follows:
1) a `CoilSupplyInputs` object is initialized, with a certain list of coils in `config.coil_names`
2) this automatically specifies what fields the container dataclass must have, so a `CoilSupplyParameter` class is defined as child of `CoilSupplyParameterABC`, and stored in the `parameter` attribute of `CoilSupplyInputs`
3) a `CoilSupplySystem` object is initialized with the `CoilSupplyInputs` object, so all methods that need to create a container can do it easily, either empty (full of `None` values) or passing an object to `validate_parameter`... dictionaries are distributed in the dataclass, other objects are just copied.

Perhaps that's not the best way to do this, but this was meant to be discussed with you and @oliverfunk 

Let me know what you think,